Fix #876 - Improve config migration by chaining together

tsyesika opened the pull request:
This implements the proposal in #876.

The migrations are now chained together so each migration is only responsible for migrating from the previous version to the next version. This means one can now simply define a migration which takes in the output of the previous migration and migrates it forward without having to modify any previous migration.

The PR also updates the 3.2.0 migration to work with the changes made in 2017.07.01 which broke the migration (due to the interfaces changing in the schema).

Each migration now should take in a string (an exception has been made for the legacy migration) and output a string. This provides a uniform interface that lets them easily be chained together (as they input and output the same things). 

I have tested the migrations both mid way through the chain and from legacy. The legacy migration I tested is located https://people.igalia.com/dpino/lwaftr/lwaftr.conf with this binidng table https://people.igalia.com/dpino/lwaftr/binding_table.txt I believe it to be 20M softwires entries.

Commit:
Fix #876 - Improve config migration by chaining together

Each migration is now only responsible for progressing to the next version
which then will run the subsequent migration to progress further. This is
done as a long chain until the configuration has been migrated to the latest
configuration schema.

This also fixes an issue with the 3.2.0 -> 2017.07.01 migration which no
longer functioned correctly.

Commit:
Add selftests for snabb lwaftr migrate-configuration

This adds two tests, one which migrates from legacy to present and the
other which migrates from 3.2.0 to present (part way through the chain).
These two migrations run the migration and verify that the migrate command
provided a succesful status code (0). It does not however check the
configs themselves.

xray7224 commented:
There were a few changes to previous migrations I had to make when looking at them. This is a side-effect of us keeping the same schema (i.e. the same name for the schema) and having no way for the migrations to use the schema it was actually written against. These changes are far less than otherwise would have been required under the old migration system though.

This PR also introduces unit testing which should flag up these problems when migrations are written. One further solution that would be trivial to do would be to include a schema for each version (if it requires one) which is what it was written against. This should reduce (hopefully to almost zero) any modifications to previous schemas. I have not done that right now, I don't feel like it's _required_ and would be an easy change to make in the future if we do desired.

wingo commented:
Looks good to me.  Please push when/if the CI goes green.

Commit:
Reduce memory footprint of migration

wingo commented on src/program/lwaftr/migrate_configuration/migrate_configuration.lua:
```diff
@@ -270,13 +270,18 @@ function load_binding_table(file)
+   temp:close()
+   conf = io.open("/tmp/migrate-configuration-temp", "r"):read("*a")
+   os.remove("/tmp/migrate-configuration-temp")
+   return conf
```
Humm, I think this one might not work -- if there are multiple migrations happening at the same time, this will lead to sadness.  Does the memory issue occur in normal operation, or just when migrating legacy configs?  If it's just when migrating legacy configs, perhaps the `collectgarbage` call below is sufficient.

tsyesika replied:
So I'm still a bit sketchy on how luajit's GC actually works but, it seemed after each migration the memory usage almost doubled. Certain migrations should cause a higher memory usage due to the nature of the config changes but not to the degree I was seeing them. It seemed that this function was largely responsible.

If I look at how it works with `memfile` if I explicitly collect the garbage where it is now so it should do a full GC collect after each migration and then print to tell me how much memory it's using before each migration it looks like this:
```
$ sudo ./snabb lwaftr migrate-configuration -f legacy lwaftr.conf
Migrating version:	legacy	(memory: 9.0424470901489)
Migrating version:	3.0.1	(memory: 462.83460521698)
Migrating version:	3.0.1.1	(memory: 430.83438396454)
Migrating version:	3.2.0	(memory: 730.37660884857)
```

config_to_string function:
```lua
local function config_to_string(schema, conf)
   if type(schema) == "string" then
      schema = yang.load_schema_by_name(schema)
   end
   local memfile = util.string_output_file()
   yang.print_config_for_schema(schema, conf, memfile)
   local migrated = memfile:flush()
   memfile = nil
   conf = nil
   return migrated
end
```

I would have expected without any of the `memfile = nil` and `config = nil` that when the function returns and the GC is explicitly called to collect then it'd free those. I go the step further and set them to nil and it still seems to balloon the memory.

What it looked like when I used the temp file writing that's currently in this PR:
```
$ sudo ./snabb lwaftr migrate-configuration -f legacy lwaftr.conf
Migrating version:	legacy	(memory: 9.0434036254883)
Migrating version:	3.0.1	(memory: 166.33897018433)
Migrating version:	3.0.1.1	(memory: 134.27617263794)

```

I don't see why in the first example after `config_to_string` has returned and presumably `memfile` can be freed it still uses 462 MB rather than the 166 MB that the second version has. I can find any tooling to do memory profiling, I'm a bit lost to why this is happening.

It seems there are two solutions:

1) Figure out why the GC isn't behaving as I think it ought to and thus how to fix it (ideal but will probably take quite a while)
2) Put something unique like exact timestamp + random token in the file name so multiple migrations can be run at once.

Any ideas?

Commit:
This fixes the memory issues with util.string_output_file

Upon futher investigation it occured to me that the `out` table used
internally in util.string_output_file might be sticking around after
due to the way it's implemented. This implements a `clear` method
on `string_output_file` that allows you to clear it once you're done.

xray7224 commented:
I actually took another look at `util.string_output_file` and it occurred to me the way it's implemented maybe the `out` table it builds up internally sticks around even when `file` has gone away. I added a method to clear it and it actually produces memory usage similar to using an external file now:

```
$ sudo ./snabb lwaftr migrate-configuration -f legacy lwaftr.conf
Migrating version:	legacy	(memory: 9.0425720214844)
Migrating version:	3.0.1	(memory: 168.23133182526)
Migrating version:	3.0.1.1	(memory: 134.73111057281)
```

Turns out it was easier than I had initially thought. PTAL @wingo, hopefully this is better :)

wingo reviewed (approved):
LGTM :)
